parser/prism/Translator.cc

```diff
@@ -18,14 +30,73 @@ std::unique_ptr<parser::Node> Translator::translate(pm_node_t *node) {
-        case PM_BLOCK_PARAMETER_NODE: {
+        case PM_BLOCK_ARGUMENT_NODE: { // A block forwarded into a method call, e.g. the `&b` in `a.map(&b)`
+            auto blockArg = reinterpret_cast<pm_block_argument_node *>(node);
+            auto loc = &blockArg->base.location;
+
+            auto expr = translate(blockArg->expression);
+
+            return make_unique<parser::BlockPass>(parser.translateLocation(loc), std::move(expr));
+        }
+        case PM_BLOCK_NODE: { // An explicit block passed to a method call, i.e. `{ ... }` or `do ... end
+            unreachable("PM_BLOCK_NODE has special handling in translateCallWithBlock, see its docs for details.");
+        }
+        case PM_BLOCK_PARAMETER_NODE: { // A block parameter declared at the top of a method, e.g. `def m(&block)`
             auto blockParamNode = reinterpret_cast<pm_block_parameter_node *>(node);
             pm_location_t *loc = &blockParamNode->base.location;
 
             std::string_view name = parser.resolveConstant(blockParamNode->name);
 
             return make_unique<parser::Blockarg>(parser.translateLocation(loc), gs.enterNameUTF8(name));
         }
+        case PM_BLOCK_PARAMETERS_NODE: { // The parameters declared at the top of a PM_BLOCK_NODE
+            auto paramsNode = reinterpret_cast<pm_block_parameters_node *>(node);
+            return translate(reinterpret_cast<pm_node *>(paramsNode->parameters));
+        }
+        case PM_CALL_NODE: {
+            auto callNode = reinterpret_cast<pm_call_node *>(node);
+            pm_location_t *loc = &callNode->base.location;
+            pm_location_t *messageLoc = &callNode->message_loc;
+
+            auto name = parser.resolveConstant(callNode->name);
+
+            std::unique_ptr<parser::Node> receiver;
+            if (auto prismReceiver = callNode->receiver; prismReceiver != nullptr) {
+                receiver = translate(prismReceiver);
+            }
+
+            absl::Span<pm_node_t *> prismArgs;
+            if (auto argsNode = callNode->arguments; argsNode != nullptr) {
+                prismArgs = absl::MakeSpan(argsNode->arguments.nodes, argsNode->arguments.size);
+            }
+
+            pm_node_t *prismBlock = callNode->block;
+            auto hasForwardedBlockArg = prismBlock != nullptr && PM_NODE_TYPE_P(prismBlock, PM_BLOCK_ARGUMENT_NODE);
+
+            parser::NodeVec args;
+            args.reserve(prismArgs.size() + (hasForwardedBlockArg ? 0 : 1));
+
+            for (auto &prismArg : prismArgs) {
+                unique_ptr<parser::Node> sorbetArg = translate(prismArg);
+                args.emplace_back(std::move(sorbetArg));
+            }
+
+            if (hasForwardedBlockArg) {
+                auto blockPassNode = translate(prismBlock);
+                args.emplace_back(std::move(blockPassNode));
+            }
+
+            auto sendNode =
+                make_unique<parser::Send>(parser.translateLocation(loc), std::move(receiver), gs.enterNameUTF8(name),
```

Comment:
This might not be UTF-8, I'm not sure if that's an issue or not.

Reply:
Yeah it's something we'll need to revisit (especially for string literals). It'll probably get baked into the `translateLocation` helper we use all over